File: go/problem_43.go
package main

import "fmt"

// MaxStack supports push, pop, and max in O(1) by maintaining a parallel max-stack.
type MaxStack struct {
	data []int
	maxs []int
}

func (s *MaxStack) Push(val int) {
	s.data = append(s.data, val)
	if len(s.maxs) == 0 || val >= s.maxs[len(s.maxs)-1] {
		s.maxs = append(s.maxs, val)
	} else {
		s.maxs = append(s.maxs, s.maxs[len(s.maxs)-1])
	}
}

func (s *MaxStack) Pop() (int, bool) {
	if len(s.data) == 0 { return 0, false }
	top := s.data[len(s.data)-1]
	s.data = s.data[:len(s.data)-1]
	s.maxs = s.maxs[:len(s.maxs)-1]
	return top, true
}

func (s *MaxStack) Max() (int, bool) {
	if len(s.maxs) == 0 { return 0, false }
	return s.maxs[len(s.maxs)-1], true
}

func problem43() {
	fmt.Println("Problem 43:")
	s := &MaxStack{}
	s.Push(1); s.Push(3); s.Push(2)
	if m, _ := s.Max(); m != 3 { panic("assertion failed") }
	if v, _ := s.Pop(); v != 2 { panic("assertion failed") }
	if m, _ := s.Max(); m != 3 { panic("assertion failed") }
	if v, _ := s.Pop(); v != 3 { panic("assertion failed") }
	if m, _ := s.Max(); m != 1 { panic("assertion failed") }
	_, ok := (&MaxStack{}).Pop()
	if ok { panic("assertion failed") }
	fmt.Println("All Tests Passed")
}

File: typescript/problem_43.ts
export class MaxStack {
    private data: number[] = [];
    private maxs: number[] = [];

    push(val: number): void {
        this.data.push(val);
        this.maxs.push(this.maxs.length === 0 ? val : Math.max(val, this.maxs[this.maxs.length - 1]));
    }

    pop(): number | null {
        if (this.data.length === 0) return null;
        this.maxs.pop();
        return this.data.pop()!;
    }

    max(): number | null {
        return this.maxs.length === 0 ? null : this.maxs[this.maxs.length - 1];
    }
}

export function problem43(): void {
    console.log("Problem 43:");
    const s = new MaxStack();
    s.push(1); s.push(3); s.push(2);
    if (s.max() !== 3) throw new Error("assertion failed");
    if (s.pop() !== 2) throw new Error("assertion failed");
    if (s.max() !== 3) throw new Error("assertion failed");
    if (s.pop() !== 3) throw new Error("assertion failed");
    if (s.max() !== 1) throw new Error("assertion failed");
    if (new MaxStack().pop() !== null) throw new Error("assertion failed");
    console.log("All Tests Passed");
}
